awkward/array/jagged.py

```diff
@@ -1578,6 +1578,8 @@ def _argminmax(self, ismin):
 
         # workaround for lack of general out[...,:1] support
         nonempty = out.counts > 0
+        if self.offsetsaliased(out._starts, out._stops):
+            out.stops = out.stops.copy()
         out.stops[nonempty] = out.starts[nonempty] + 1
```

Comment:
I guess the usage of `__getitem__` on the return of `out.stops` precludes the ability to hook the change and respond appropriately.  I would argue that this means either the `stops` array should have some write protection flag set, or that `JaggedArray.stops` should not be a public member.

Reply:
It has been growing clear that all of the array properties, like `starts` and `stops` on JaggedArray and `mask` on MaskedArray, etc., are in a middle ground between public and private.

They're not entirely private like the methods that start with an underscore, which really are implementation details that other libraries shouldn't depend on. (I've run into a few mistakes of this sort where uproot has depended on a method that starts with underscore, with disastrous consequences.) These array properties _are_ part of the public API that libraries like uproot can and should depend on.

On the other hand, they're too low-level to be regularly manipulated in a data analysis. There's a set of things to worry about, like multidimensional starts and stops (the error you ran into), aliasing (the one I ran into), unreachable elements (which totally confused the students at the HATS), etc. It's reasonable for programmers to be expected to maintain this contract, but not data analysts, at least not _while_ they're also trying to focus on data analysis.

That's precisely why these structure classes and their properties will be partially hidden in awkward 1.0 (see "motivation and requirements" document in [awkward-1.0](https://github.com/jpivarski/awkward-1.0)). Data analysts would see a single `Array` class, and if they choose, they can get to the underlying JaggedArray/MaskedArray/etc. through an `array.struct` field, which will have all the JaggedArrays with public `starts`, `stops`, etc.

Also, JaggedArray is going to be split into two classes, one that is based on `offsets` and the other based on `starts` and `stops` with no aliasing. Some methods on the JaggedArray with `offsets` return a JaggedArray with `starts` and `stops` ("JaggedArray vs RaggedArray?" "ListArray vs ListOffsetsArray"? not sure of names yet). Hiding the structure under a single `Array` class lets us introduce more structure classes because ordinary users won't have to understand their differences. I'll also be removing multidimensional `starts` and `stops` and moving this functionality to a new structure class.